### PR TITLE
Improve nocloud detection / improve logging

### PR DESCRIPTION
**Pull request by holmanb**
See individual commits, relates to https://github.com/canonical/cloud-init/issues/5389

### CODE REVIEW

File: cloudinit/sources/DataSourceNoCloud.py [hunk 1]
```diff
@@ -408,6 +436,8 @@ def ds_detect(self):
             if serial == "nocloud-net":
                 log_deprecated()
             return True
+        elif self.sys_cfg.get("datasource", {}).get("NoCloud").get("seedfrom"):
```
Comment (aciba90) [inline cloudinit/sources/DataSourceNoCloud.py:439]: Do you think it is worth {unit}testing this?

**Review by aciba90 — COMMENTED** [review]
Thanks for tackling this!

I disagree with https://github.com/canonical/cloud-init/pull/5515/commits/c2e6508eff6f21446f8a963e18d02fdba0f85285.

The progress is logged using `events.ReportEventStack` which is logged from `handlers.py` module, see the attached log.

IMHO, many of the log additions are either:

1. Already covered by a ReportEventStack and feel redundant
2. Implementation details of cloud-init that users are not aware of
3. The log message are that generic, that provide no context without reading the source code

I would be careful adding even more logs to the already busy log file.

```
2024-07-17 10:02:41,262 - handlers.py[DEBUG]: start: init-local/check-cache: attempting to read from cache [check]
2024-07-17 10:02:41,272 - handlers.py[DEBUG]: finish: init-local/check-cache: SUCCESS: restored from checked cache: DataSourceLXD
2024-07-17 10:02:41,300 - handlers.py[DEBUG]: finish: init-local: SUCCESS: searching for local datasources
2024-07-17 10:02:41,681 - handlers.py[DEBUG]: start: init-network/check-cache: attempting to read from cache [trust]
2024-07-17 10:02:41,683 - handlers.py[DEBUG]: finish: init-network/check-cache: SUCCESS: restored from cache with run check: DataSourceLXD
2024-07-17 10:02:41,705 - handlers.py[DEBUG]: start: init-network/setup-datasource: setting up datasource
2024-07-17 10:02:41,705 - handlers.py[DEBUG]: finish: init-network/setup-datasource: SUCCESS: setting up datasource
2024-07-17 10:02:41,708 - handlers.py[DEBUG]: start: init-network/consume-user-data: reading and applying user-data
2024-07-17 10:02:41,710 - handlers.py[DEBUG]: finish: init-network/consume-user-data: SUCCESS: reading and applying user-data
2024-07-17 10:02:41,710 - handlers.py[DEBUG]: start: init-network/consume-vendor-data: reading and applying vendor-data
2024-07-17 10:02:41,710 - handlers.py[DEBUG]: finish: init-network/consume-vendor-data: SUCCESS: reading and applying vendor-data
2024-07-17 10:02:41,710 - handlers.py[DEBUG]: start: init-network/consume-vendor-data2: reading and applying vendor-data2
2024-07-17 10:02:41,710 - handlers.py[DEBUG]: finish: init-network/consume-vendor-data2: SUCCESS: reading and applying vendor-data2
2024-07-17 10:02:41,721 - handlers.py[DEBUG]: start: init-network/activate-datasource: activating datasource
2024-07-17 10:02:41,722 - handlers.py[DEBUG]: finish: init-network/activate-datasource: SUCCESS: activating datasource
2024-07-17 10:02:41,758 - handlers.py[DEBUG]: start: init-network/config-seed_random: running config-seed_random with frequency once-per-instance
2024-07-17 10:02:41,758 - handlers.py[DEBUG]: finish: init-network/config-seed_random: SUCCESS: config-seed_random previously ran
2024-07-17 10:02:41,758 - handlers.py[DEBUG]: start: init-network/config-growpart: running config-growpart with frequency always
2024-07-17 10:02:41,763 - handlers.py[DEBUG]: finish: init-network/config-growpart: SUCCESS: config-growpart ran successfully
2024-07-17 10:02:41,763 - handlers.py[DEBUG]: start: init-network/config-resizefs: running config-resizefs with frequency always
2024-07-17 10:02:41,764 - handlers.py[DEBUG]: finish: init-network/config-resizefs: SUCCESS: config-resizefs ran successfully
2024-07-17 10:02:41,764 - handlers.py[DEBUG]: start: init-network/config-mounts: running config-mounts with frequency once-per-instance
2024-07-17 10:02:41,765 - handlers.py[DEBUG]: finish: init-network/config-mounts: SUCCESS: config-mounts previously ran
2024-07-17 10:02:41,765 - handlers.py[DEBUG]: start: init-network/config-set_hostname: running config-set_hostname with frequency once-per-instance
2024-07-17 10:02:41,765 - handlers.py[DEBUG]: finish: init-network/config-set_hostname: SUCCESS: config-set_hostname previously ran
2024-07-17 10:02:41,765 - handlers.py[DEBUG]: start: init-network/config-update_hostname: running config-update_hostname with frequency always
2024-07-17 10:02:41,765 - handlers.py[DEBUG]: finish: init-network/config-update_hostname: SUCCESS: config-update_hostname ran successfully
2024-07-17 10:02:41,766 - handlers.py[DEBUG]: start: init-network/config-users_groups: running config-users_groups with frequency once-per-instance
2024-07-17 10:02:41,766 - handlers.py[DEBUG]: finish: init-network/config-users_groups: SUCCESS: config-users_groups previously ran
2024-07-17 10:02:41,766 - handlers.py[DEBUG]: start: init-network/config-ssh: running config-ssh with frequency once-per-instance
2024-07-17 10:02:41,766 - handlers.py[DEBUG]: finish: init-network/config-ssh: SUCCESS: config-ssh previously ran
2024-07-17 10:02:41,767 - handlers.py[DEBUG]: finish: init-network: SUCCESS: searching for network datasources
2024-07-17 10:02:42,177 - handlers.py[DEBUG]: start: modules-config/config-ssh_import_id: running config-ssh_import_id with frequency once-per-instance
2024-07-17 10:02:42,177 - handlers.py[DEBUG]: finish: modules-config/config-ssh_import_id: SUCCESS: config-ssh_import_id previously ran
2024-07-17 10:02:42,178 - handlers.py[DEBUG]: start: modules-config/config-locale: running config-locale with frequency once-per-instance
2024-07-17 10:02:42,178 - handlers.py[DEBUG]: finish: modules-config/config-locale: SUCCESS: config-locale previously ran
2024-07-17 10:02:42,178 - handlers.py[DEBUG]: start: modules-config/config-set_passwords: running config-set_passwords with frequency once-per-instance
2024-07-17 10:02:42,178 - handlers.py[DEBUG]: finish: modules-config/config-set_passwords: SUCCESS: config-set_passwords previously ran
2024-07-17 10:02:42,178 - handlers.py[DEBUG]: start: modules-config/config-grub_dpkg: running config-grub_dpkg with frequency once-per-instance
2024-07-17 10:02:42,178 - handlers.py[DEBUG]: finish: modules-config/config-grub_dpkg: SUCCESS: config-grub_dpkg previously ran
2024-07-17 10:02:42,178 - handlers.py[DEBUG]: start: modules-config/config-apt_configure: running config-apt_configure with frequency once-per-instance
2024-07-17 10:02:42,179 - handlers.py[DEBUG]: finish: modules-config/config-apt_configure: SUCCESS: config-apt_configure previously ran
2024-07-17 10:02:42,179 - handlers.py[DEBUG]: start: modules-config/config-byobu: running config-byobu with frequency once-per-instance
2024-07-17 10:02:42,179 - handlers.py[DEBUG]: finish: modules-config/config-byobu: SUCCESS: config-byobu previously ran
2024-07-17 10:02:42,180 - handlers.py[DEBUG]: finish: modules-config: SUCCESS: running modules for config
2024-07-17 10:02:42,830 - handlers.py[DEBUG]: start: modules-final/config-reset_rmc: running config-reset_rmc with frequency once-per-instance
2024-07-17 10:02:42,830 - handlers.py[DEBUG]: finish: modules-final/config-reset_rmc: SUCCESS: config-reset_rmc previously ran
2024-07-17 10:02:42,831 - handlers.py[DEBUG]: start: modules-final/config-scripts_vendor: running config-scripts_vendor with frequency once-per-instance
2024-07-17 10:02:42,831 - handlers.py[DEBUG]: finish: modules-final/config-scripts_vendor: SUCCESS: config-scripts_vendor previously ran
2024-07-17 10:02:42,831 - handlers.py[DEBUG]: start: modules-final/config-scripts_per_once: running config-scripts_per_once with frequency once
2024-07-17 10:02:42,831 - handlers.py[DEBUG]: finish: modules-final/config-scripts_per_once: SUCCESS: config-scripts_per_once previously ran
2024-07-17 10:02:42,831 - handlers.py[DEBUG]: start: modules-final/config-scripts_per_boot: running config-scripts_per_boot with frequency always
2024-07-17 10:02:42,831 - handlers.py[DEBUG]: finish: modules-final/config-scripts_per_boot: SUCCESS: config-scripts_per_boot ran successfully
2024-07-17 10:02:42,831 - handlers.py[DEBUG]: start: modules-final/config-scripts_per_instance: running config-scripts_per_instance with frequency once-per-instance
2024-07-17 10:02:42,831 - handlers.py[DEBUG]: finish: modules-final/config-scripts_per_instance: SUCCESS: config-scripts_per_instance previously ran
2024-07-17 10:02:42,832 - handlers.py[DEBUG]: start: modules-final/config-scripts_user: running config-scripts_user with frequency once-per-instance
2024-07-17 10:02:42,832 - handlers.py[DEBUG]: finish: modules-final/config-scripts_user: SUCCESS: config-scripts_user previously ran
2024-07-17 10:02:42,832 - handlers.py[DEBUG]: start: modules-final/config-ssh_authkey_fingerprints: running config-ssh_authkey_fingerprints with frequency once-per-instance
2024-07-17 10:02:42,832 - handlers.py[DEBUG]: finish: modules-final/config-ssh_authkey_fingerprints: SUCCESS: config-ssh_authkey_fingerprints previously ran
2024-07-17 10:02:42,832 - handlers.py[DEBUG]: start: modules-final/config-keys_to_console: running config-keys_to_console with frequency once-per-instance
2024-07-17 10:02:42,832 - handlers.py[DEBUG]: finish: modules-final/config-keys_to_console: SUCCESS: config-keys_to_console previously ran
2024-07-17 10:02:42,832 - handlers.py[DEBUG]: start: modules-final/config-install_hotplug: running config-install_hotplug with frequency once-per-instance
2024-07-17 10:02:42,832 - handlers.py[DEBUG]: finish: modules-final/config-install_hotplug: SUCCESS: config-install_hotplug previously ran
2024-07-17 10:02:42,833 - handlers.py[DEBUG]: start: modules-final/config-final_message: running config-final_message with frequency always
2024-07-17 10:02:42,837 - handlers.py[DEBUG]: finish: modules-final/config-final_message: SUCCESS: config-final_message ran successfully
2024-07-17 10:02:42,838 - handlers.py[DEBUG]: finish: modules-final: SUCCESS: running modules for final
```

**Comment by holmanb** [comment]
> I disagree with https://github.com/canonical/cloud-init/commit/c2e6508eff6f21446f8a963e18d02fdba0f85285.

I still think that we can improve on current state. The majority of those events report on specific cloud-config modules running, which is still very low in the call stack. Also, having all progress reported from a single source file named `handlers.py` is confusing because 1) searching the docs for "handlers" will find many references to part-handlers and 2) the output has a lot of noise and redundancy - not very human friendly.

You're right though, that progress logging improvements should use existing infrastructure rather than an alternative to avoid redundancy.

I've dropped that commit for now and will rethink the approach.

File: cloudinit/sources/DataSourceNoCloud.py (hunk 1 above)
Comment (holmanb) [reply cloudinit/sources/DataSourceNoCloud.py:445]: Done

File: tools/ds-identify [hunk 2]
```diff
@@ -989,8 +989,12 @@ dscheck_NoCloud() {
     fi
 
     # This is a bit hacky, but a NoCloud false positive isn't the end of the world
-    if check_config "NoCloud" && check_config "user-data" && check_config "meta-data"; then
-        return ${DS_FOUND}
+    if check_config "NoCloud"; then
+        if check_config "user-data" && check_config "meta-data"; then
+            return ${DS_FOUND}
+        elif check_config "seedfrom"; then
```
Comment (aciba90) [inline tools/ds-identify:995]: `DataSourceOVF` does also use `seedfrom`, as `check_config` does effectively a grep-search, should check for the abscence of OVF in sys-config files?

```suggestion
        elif ! check_config "OVF" && check_config "seedfrom"; then
```

File: tests/unittests/test_ds_identify.py
Comment (aciba90) [inline tests/unittests/test_ds_identify.py:1]: Cool tests, thanks!

File: cloudinit/sources/DataSourceNoCloud.py [hunk 3]
```diff
@@ -121,6 +130,12 @@ def _pp2d_callback(mp, data):
 
         label = self.ds_cfg.get("fs_label", "cidata")
         if label is not None:
+            if label.lower() != "cidata":
+                util.deprecate(
+                    deprecated="Custom fs_label keys",
+                    deprecated_version="22.3",
```
Comment (aciba90) [inline cloudinit/sources/DataSourceNoCloud.py:136]: Why 22.3? Do you mean 24.3?

Why deprecate it? I see:

1. https://github.com/canonical/cloud-init/issues/3694#issuecomment-1545255090
2. https://github.com/canonical/cloud-init/issues/5062#issuecomment-2066124987

1) The creator propose the deprecation and 2) lays out an arguably very corner-casey user-case that this feature solves.

**Review by aciba90 — CHANGES_REQUESTED** [review]
Thanks, @holmanb, for adding unit-tests and extending `ds-identify` too. I left some in-line ask-for-clarification / requests comments.

> I still think that we can improve on current state. The majority of those events report on specific cloud-config modules running, which is still very low in the call stack. Also, having all progress reported from a single source file named handlers.py is confusing because 1) searching the docs for "handlers" will find many references to part-handlers and 2) the output has a lot of noise and redundancy - not very human friendly.

> You're right though, that progress logging improvements should use existing infrastructure rather than an alternative to avoid redundancy.

> I've dropped that commit for now and will rethink the approach.

I agree, could we move this discussion to another issue / PR?

File: tools/ds-identify (hunk 2 above)
Comment (holmanb) [reply tools/ds-identify:995]: Good point, that seems reasonable.

Comment (holmanb) [reply tools/ds-identify:995]: @aciba90 actually, no I don't think that doing this makes sense. The seedfrom used by DataSourceOVF comes through the xml file, not system configuration.

File: cloudinit/sources/DataSourceNoCloud.py (hunk 3 above)
Comment (dermotbradley) [reply cloudinit/sources/DataSourceNoCloud.py:136]: I seem to recall being able to change (after 1st boot) the NoCloud label checked for as being mentioned as a way to deal with potential security issues where cloud-init is used on **physical** machines to address the risk that someone could, some time after 1st boot, connect a USB device with a "CIDATA" label to manipulate the OS.

I don't remember where I saw that mentioned, however I did find a related Ubuntu and cloud-init issue: 

https://ubuntu.com/security/CVE-2020-11933
https://bugs.launchpad.net/snapd/+bug/1879530

Comment (aciba90) [reply cloudinit/sources/DataSourceNoCloud.py]: Agreed, let's deprecate!

**Review by aciba90 — APPROVED** [review]
LGTM, thanks!

Black complains about a change in a file that wasn't touched in this branch.

File: cloudinit/sources/DataSourceNoCloud.py (hunk 3 above)
Comment (holmanb) [reply cloudinit/sources/DataSourceNoCloud.py]: Thanks for input @dermotbradley!

> I seem to recall being able to change (after 1st boot) the NoCloud label checked for as being mentioned in the past as a way to deal with potential security issues where cloud-init is used on physical machines to address the risk that someone could, some time after 1st boot, connect a USB device with a "CIDATA" label to manipulate the OS.

The CVE targeted snapd, not cloud-init. Snapd/snap core resolved this by disabling cloud-init on first boot.

That said, security by obscurity using a low-entropy label (limited length, ascii-only) sounds dubious, and better mitigations exist for most relevant use cases (disabling cloud-init, forcibly use DataSourceNone instead of NoCloud for local configuration).

The ability to require signed configuration data would be a huge win for anybody that currently depends on using this mitigation, which I think someone else is currently working on. That should provide stronger mitigations than a custom label currently does for the use cases that I didn't already mention above: one wants to be able to use an attachable disk but not allow others to, or someone wants to pull user-data from a URI - both of which would be difficult to secure for a myriad of reasons when attackers are expected to have physical access to a system. 

Since we're deprecating and not removing this, the deprecation cycle time should give ample time for anybody who is using this to raise concern about specific use cases - as well as for us to release better alternatives (or to change our mind about the deprecation if better alternatives cannot be provided).

**Comment by holmanb** [comment]
> LGTM, thanks!
> 
> Black complains about a change in a file that wasn't touched in this branch.

Thanks for the review @aciba90! I just rebased on main which should resolve that issue.
